ENH: Config module to replace long list of downstreamed arguments

oesteban opened the pull request:
~~Includes addressing #2019, although~~ I think that one could be addressed separately to simplify this already huge PR.

Requires: https://github.com/nipreps/niworkflows/pull/475

## Summary
This module implements the memory structures to keep a consistent, singleton config.

### Example
```Python
>>> from fmriprep import config
>>> config_file = config.execution.work_dir / '.fmriprep.toml'
>>> config.to_filename(config_file)
>>> # Call build_workflow(config_file, retval) in a subprocess
>>> with Manager() as mgr:
>>>     from .workflow import build_workflow
>>>     retval = mgr.dict()
>>>     p = Process(target=build_workflow, args=(str(config_file), retval))
>>>     p.start()
>>>     p.join()
>>> config.load(config_file)
>>> # Access configs from any code section as:
>>> value = config.section.setting
```
The module also has a :py:func:`to_filename` function to allow writting out
the settings to hard disk in *ToML* format, which looks like:

    [nipype]
    crashfile_format = "txt"
    get_linked_libs = false
    nprocs = 8
    omp_nthreads = 8
    plugin = "MultiProc"
    resource_monitor = false
    stop_on_first_crash = false
    version = "1.5.0-dev+gdffcb7815"

    [execution]
    bids_dir = "/data/openfmri/ds000005"
    boilerplate_only = false
    exec_env = "posix"
    fs_license_file = "/opt/freesurfer/license.txt"
    fs_subjects_dir = "/data/openfmri/ds000005/derivatives/freesurfer-6.0.1"
    log_dir = "/data/openfmri/ds000005/derivatives/fmriprep/logs"
    log_level = 15
    low_mem = false
    md_only_boilerplate = false
    notrack = true
    output_dir = "/data/openfmri/ds000005/derivatives"
    reports_only = false
    run_uuid = "20200302-174345_9ba9f304-82de-4538-8c3a-570c5f5d8f2f"
    participant_label = [ "01",]
    version = "20.0.1+13.g3ca42930.dirty"
    work_dir = "/home/oesteban/tmp/fmriprep-ds005/fprep-work2"
    write_graph = false

    [workflow]
    anat_only = false
    aroma_err_on_warn = false
    aroma_melodic_dim = -200
    bold2t1w_dof = 6
    cifti_output = false
    fmap_bspline = false
    force_syn = false
    hires = true
    ignore = []
    internal_spaces = "MNI152NLin2009cAsym"
    longitudinal = false
    medial_surface_nan = false
    run_reconall = true
    skull_strip_fixed_seed = false
    skull_strip_template = "OASIS30ANTs"
    t2s_coreg = false
    use_aroma = false

    [nipype.plugin_args]
    maxtasksperchild = 1
    raise_insufficient = false

This config file is used to pass the settings across processes,
using the :py:func:`load` function.

Other responsibilities of the config module:

  * Switching Python's ``multiprocessing`` to *forkserver* mode.
  * Set up new logger levels (25: IMPORTANT, and 15: VERBOSE).
  * Set up a warnings filter as soon as possible.
  * Initialize runtime descriptive settings (e.g., default FreeSurfer license,
    execution environment, nipype and *fMRIPrep* versions, etc.).
  * Automated I/O magic operations:

    * :obj:`Path` \<-\> :obj:`str` \<-\> :obj:`Path`).
    * :py:class:`~niworkflows.util.spaces.SpatialReferences` \<-\> :obj:`str` \<-\>
      :py:class:`~niworkflows.util.spaces.SpatialReferences`
    * :py:class:`~bids.layout.BIDSLayout` \<-\> :obj:`str` \<-\>
      :py:class:`~bids.layout.BIDSLayout`

### Potential utilizations of the config module
- Crawling versions of relevant software and storing them in the read-only section (`environment`) of the config file, so that those are then reported in the log folder and within sentry.
- Keeping track of random seeds (generating them and keeping them for future execution).

effigies commented:
@oesteban Here's the merge conflict between this and 20.0.2:

```
def main():
    """Entry point."""
    import os
    import sys
    import gc
    from nipype import config as ncfg
    from multiprocessing import Process, Manager
    from .parser import parse_args
    from ..utils.bids import write_derivative_description

    parse_args()

    sentry_sdk = None
    if not config.execution.notrack:
        import sentry_sdk
        from ..utils.sentry import sentry_setup
        sentry_setup()

    config_file = config.execution.work_dir / '.fmriprep.toml'
    config.to_filename(config_file)

<<<<<<< HEAD
    # Precedence: --fs-license-file, $FS_LICENSE, default_license
    license_file = opts.fs_license_file

    if license_file is None and os.getenv("FS_LICENSE"):
        license_file = Path(os.getenv("FS_LICENSE"))

    if license_file is None and os.getenv("FREESURFER_HOME"):
        license_file = Path(os.getenv("FREESURFER_HOME")) / "license.txt"

    # FreeSurfer license
    if license_file is None or not license_file.exists():
        raise RuntimeError("""\
ERROR: a valid license file is required for FreeSurfer to run. fMRIPrep looked for an existing \
license file at several paths, in this order: 1) command line argument ``--fs-license-file``; \
2) ``$FS_LICENSE`` environment variable; and 3) the ``$FREESURFER_HOME/license.txt`` path. Get it \
(for free) by registering at https://surfer.nmr.mgh.harvard.edu/registration.html""")
    os.environ["FS_LICENSE"] = str(license_file.absolute())

    # Retrieve logging level
    log_level = int(max(25 - 5 * opts.verbose_count, logging.DEBUG))
    # Set logging
    logger.setLevel(log_level)
    logger.addHandler(logging.StreamHandler())
    nlogging.getLogger('nipype.workflow').setLevel(log_level)
    nlogging.getLogger('nipype.interface').setLevel(log_level)
    nlogging.getLogger('nipype.utils').setLevel(log_level)

    # Call build_workflow(opts, retval)
=======
    # Call build_workflow(config_file, retval) in a subprocess
>>>>>>> oesteban/maint/config-file
    with Manager() as mgr:
        from .workflow import build_workflow
        retval = mgr.dict()
        p = Process(target=build_workflow, args=(str(config_file), retval))
        p.start()
        p.join()

        retcode = p.exitcode or retval.get('return_code', 0)
        fmriprep_wf = retval.get('workflow', None)

    config.load(config_file)
```

Would you prefer to resolve that in this PR, or after this is done? (Trying to decide whether to merge the latest tag into `master`.)

pep8speaks commented:
Hello @oesteban! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





* In the file [`fmriprep/cli/parser.py`](https://github.com/poldracklab/fmriprep/blob/d266b7847c5fb9f41b1fd616eb3a05119d4bdaa6/fmriprep/cli/parser.py):

> [Line 54:100](https://github.com/poldracklab/fmriprep/blob/d266b7847c5fb9f41b1fd616eb3a05119d4bdaa6/fmriprep/cli/parser.py#L54): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 99 characters)

oesteban commented:
> Would you prefer to resolve that in this PR, or after this is done?

I think after

oesteban commented:
Okay, after e7d7ac81dff8e055c7beed85e85a2af01f431156 this might be comming to a reviewable state.

- [x] A new `mock_config` context helps with documentation building / testing. It includes a stub with one subject from ds005 (i.e., all the boilerplate to generate fake data within the workflow creation methods is gone).
- [x] I have deferred the `BIDSLayout`'s `database_path` issue for the future.
- [ ] An interesting sequel of this PR would include crawling versions of relevant software and storing them in the read-only section (`environment`) of the config file, so that those are then reported in the log folder and within sentry.
- [ ] Should I just get rid of pytest-putty? (see travis failing).

Docs are built here - I'm going to add the new config module to the API
https://4979-53608443-gh.circle-artifacts.com/0/tmp/src/fmriprep/docs/_build/html/index.html

effigies commented:
Yes, we can drop pytest putty.

oesteban commented on fmriprep/cli/tests/test_parser.py:
```diff
@@ -0,0 +1,116 @@
+
+    assert ('FLAGGED' in captured) is flagged[0]
+    if flagged[0]:
+        assert ((flagged[1] or 'reason: unknown') in captured)
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/cli/tests/test_parser.py:
```diff
@@ -0,0 +1,116 @@
+        + ['--mem', argval]
+    opts = _build_parser().parse_args(args)
+
+    assert opts.memory_gb == gb
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/cli/tests/test_parser.py:
```diff
@@ -0,0 +1,116 @@
+
+    opts = _build_parser().parse_args(args)
+
+    assert opts.bids_dir == datapath
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/cli/tests/test_parser.py:
```diff
@@ -0,0 +1,116 @@
+Please check out our documentation about how and when to upgrade:
+https://fmriprep.readthedocs.io/en/latest/faq.html#upgrading""" % (current, latest)
+
+    assert (msg in captured) is expectation
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/config.py:
```diff
@@ -1,5 +1,556 @@
+                _oc_limit = '{}%'.format(
+                    Path('/proc/sys/vm/overcommit_ratio').read_text().strip()
+                )
+except Exception:
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Try, Except, Pass detected.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Try, Except, Pass detected.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Try, Except, Pass detected.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/cli/workflow.py:
```diff
@@ -0,0 +1,166 @@
+    }
+
+    if not config.execution.md_only_boilerplate and citation_files['md'].exists():
+        from subprocess import check_call, CalledProcessError, TimeoutExpired
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Consider possible security implications associated with check_call module.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Consider possible security implications associated with check_call module.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Consider possible security implications associated with check_call module.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Consider possible security implications associated with check_call module.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Consider possible security implications associated with check_call module.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Consider possible security implications associated with check_call module.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Consider possible security implications associated with check_call module.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Consider possible security implications associated with check_call module.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Consider possible security implications associated with check_call module.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/cli/tests/test_parser.py:
```diff
@@ -0,0 +1,116 @@
+    with pytest.raises(SystemExit) as error:
+        _build_parser().parse_args(args)
+
+    assert error.value.code == code
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/cli/workflow.py:
```diff
@@ -0,0 +1,166 @@
+        config.loggers.cli.info(
+            'Generating a LaTeX version of the citation boilerplate...')
+        try:
+            check_call(cmd, timeout=10)
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/cli/tests/test_parser.py:
```diff
@@ -0,0 +1,116 @@
+    _build_parser()
+    captured = capsys.readouterr().err
+
+    assert ('FLAGGED' in captured) is flagged[0]
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/tests/test_config.py:
```diff
@@ -0,0 +1,28 @@
+    assert "MNI152NLin6Asym:res-2" in [
+        str(s) for s in spaces.get_standard(full_spec=True)]
+
+    assert "MNI152NLin6Asym_res-2" in [
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/tests/test_config.py:
```diff
@@ -0,0 +1,28 @@
+    config.init_spaces()
+
+    spaces = config.workflow.spaces
+    assert "MNI152NLin6Asym:res-2" in [
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/tests/test_config.py:
```diff
@@ -0,0 +1,28 @@
+    assert "MNI152NLin6Asym:res-2" in [
+        str(s) for s in spaces.get_standard(full_spec=True)]
+
+    assert "MNI152NLin6Asym_res-2" in [
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/config.py:
```diff
@@ -1,5 +1,607 @@
+                _oc_limit = '{}%'.format(
+                    Path('/proc/sys/vm/overcommit_ratio').read_text().strip()
+                )
+except Exception:
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Try, Except, Pass detected.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/tests/test_config.py:
```diff
@@ -0,0 +1,28 @@
+    config.init_spaces()
+
+    spaces = config.workflow.spaces
+    assert "MNI152NLin6Asym:res-2" in [
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/config.py:
```diff
@@ -1,5 +1,608 @@
+                _oc_limit = '{}%'.format(
+                    Path('/proc/sys/vm/overcommit_ratio').read_text().strip()
+                )
+except Exception:
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Try, Except, Pass detected.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/config.py:
```diff
@@ -1,5 +1,652 @@
+                _oc_limit = '{}%'.format(
+                    Path('/proc/sys/vm/overcommit_ratio').read_text().strip()
+                )
+except Exception:
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Try, Except, Pass detected.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Try, Except, Pass detected.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Try, Except, Pass detected.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Try, Except, Pass detected.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/tests/test_config.py:
```diff
@@ -0,0 +1,40 @@
+    config.init_spaces()
+    spaces = config.workflow.spaces
+
+    assert "MNI152NLin6Asym:res-2" in [
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/tests/test_config.py:
```diff
@@ -0,0 +1,40 @@
+    assert "MNI152NLin6Asym:res-2" in [
+        str(s) for s in spaces.get_standard(full_spec=True)]
+
+    assert "MNI152NLin6Asym_res-2" in [
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/tests/test_config.py:
```diff
@@ -0,0 +1,40 @@
+    assert "MNI152NLin6Asym:res-2" not in [
+        str(s) for s in spaces.get_standard(full_spec=True)]
+
+    assert "MNI152NLin6Asym_res-2" not in [
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/tests/test_config.py:
```diff
@@ -0,0 +1,40 @@
+    config.init_spaces()
+
+    spaces = config.workflow.spaces
+    assert "MNI152NLin6Asym:res-2" not in [
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented:
All greens - let's end this one @mgxd @effigies.

One thing you'll probably want to change is section names. Particularly, replacing `config.workflow` with `config.user` or similar.

mgxd commented on fmriprep/cli/parser.py:
```diff
@@ -0,0 +1,405 @@
+    def _path_exists(path, parser):
+        """Ensure a given path exists."""
+        if path is None or not Path(path).exists():
+            raise parser.error("Path does not exist: <%s>." % path)
```
```suggestion
            raise parser.error(f"Path does not exist: <{path}>.")
```

mgxd commented on fmriprep/cli/parser.py:
```diff
@@ -0,0 +1,405 @@
+        if value and Path(value).exists():
+            return loads(Path(value).read_text())
+
+    verstr = 'fMRIPrep v{}'.format(config.environment.version)
```
Now that we're 3.6+, we should prioritize fstrings IMO

```suggestion
    verstr = f'fMRIPrep v{config.environment.version}'
```

oesteban replied:
Agreed - but in a different, ad-hoc PR.

mgxd commented on fmriprep/cli/parser.py:
```diff
@@ -0,0 +1,405 @@
+
+    # Ensure input and output folders are not the same
+    if output_dir == bids_dir:
+        config.loggers.cli.error(
```
I mistakenly changed these, they should raise `parser.error`s

```suggestion
        raise parser.error(
```

mgxd commented on fmriprep/cli/parser.py:
```diff
@@ -0,0 +1,405 @@
+        )
+
+    if bids_dir in work_dir.parents:
+        config.loggers.cli.error(
```
```suggestion
        raise parser.error(
```

oesteban commented on fmriprep/cli/workflow.py:
```diff
@@ -0,0 +1,165 @@
+        config.loggers.cli.info(
+            'Generating a LaTeX version of the citation boilerplate...')
+        try:
+            check_call(cmd, timeout=10)
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/cli/workflow.py:
```diff
@@ -0,0 +1,165 @@
+    }
+
+    if not config.execution.md_only_boilerplate and citation_files['md'].exists():
+        from subprocess import check_call, CalledProcessError, TimeoutExpired
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Consider possible security implications associated with check_call module.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Consider possible security implications associated with check_call module.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Consider possible security implications associated with check_call module.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/tests/test_config.py:
```diff
@@ -0,0 +1,41 @@
+    assert "MNI152NLin6Asym:res-2" not in [
+        str(s) for s in spaces.get_standard(full_spec=True)]
+
+    assert "MNI152NLin6Asym_res-2" not in [
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/tests/test_config.py:
```diff
@@ -0,0 +1,41 @@
+    config.init_spaces()
+
+    spaces = config.workflow.spaces
+    assert "MNI152NLin6Asym:res-2" not in [
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/tests/test_config.py:
```diff
@@ -0,0 +1,41 @@
+    assert "MNI152NLin6Asym:res-2" in [
+        str(s) for s in spaces.get_standard(full_spec=True)]
+
+    assert "MNI152NLin6Asym_res-2" in [
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/tests/test_config.py:
```diff
@@ -0,0 +1,41 @@
+    config.init_spaces()
+    spaces = config.workflow.spaces
+
+    assert "MNI152NLin6Asym:res-2" in [
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/config.py:
```diff
@@ -1,5 +1,646 @@
+                _oc_limit = '{}%'.format(
+                    Path('/proc/sys/vm/overcommit_ratio').read_text().strip()
+                )
+except Exception:
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Try, Except, Pass detected.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/config.py:
```diff
@@ -1,5 +1,591 @@
+                _oc_limit = '{}%'.format(
+                    Path('/proc/sys/vm/overcommit_ratio').read_text().strip()
+                )
+except Exception:
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Try, Except, Pass detected.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Try, Except, Pass detected.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Try, Except, Pass detected.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

mgxd commented on fmriprep/cli/workflow.py:
```diff
@@ -0,0 +1,165 @@
+    config.load(config_file)
+    logs_path = config.execution.output_dir / 'fmriprep' / 'logs'
+    boilerplate = workflow.visit_desc()
+
```
```suggestion
    citation_files = {
            ext: logs_path / ('CITATION.%s' % ext)
            for ext in ('bib', 'tex', 'md', 'html')
    }
```

mgxd commented on fmriprep/cli/workflow.py:
```diff
@@ -0,0 +1,165 @@
+        citation_files = {
+            ext: logs_path / ('CITATION.%s' % ext)
+            for ext in ('bib', 'tex', 'md', 'html')
+        }
```
```suggestion
```

mgxd commented on fmriprep/cli/workflow.py:
```diff
@@ -0,0 +1,165 @@
+    citation_files = {
+        ext: config.execution.output_dir / 'fmriprep' / 'logs'
+        / ('CITATION.%s' % ext) for ext in ('bib', 'tex', 'md', 'html')
+    }
```
```suggestion
```

oesteban commented on fmriprep/tests/test_config.py:
```diff
@@ -0,0 +1,53 @@
+    config.init_spaces()
+    spaces = config.workflow.spaces
+
+    assert [str(s) for s in spaces.get_standard(full_spec=True)] == []
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/tests/test_config.py:
```diff
@@ -0,0 +1,53 @@
+    assert "MNI152NLin6Asym:res-2" not in [
+        str(s) for s in spaces.get_standard(full_spec=True)]
+
+    assert "MNI152NLin6Asym_res-2" not in [
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/tests/test_config.py:
```diff
@@ -0,0 +1,53 @@
+    config.init_spaces()
+    spaces = config.workflow.spaces
+
+    assert "MNI152NLin6Asym:res-2" in [
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/tests/test_config.py:
```diff
@@ -0,0 +1,53 @@
+    config.init_spaces()
+
+    spaces = config.workflow.spaces
+    assert "MNI152NLin6Asym:res-2" not in [
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/tests/test_config.py:
```diff
@@ -0,0 +1,53 @@
+    assert "MNI152NLin6Asym:res-2" in [
+        str(s) for s in spaces.get_standard(full_spec=True)]
+
+    assert "MNI152NLin6Asym_res-2" in [
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/cli/workflow.py:
```diff
@@ -0,0 +1,160 @@
+        config.loggers.cli.info(
+            'Generating a LaTeX version of the citation boilerplate...')
+        try:
+            check_call(cmd, timeout=10)
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/cli/parser.py:
```diff
@@ -0,0 +1,411 @@
+    )
+    from packaging.version import Version
+    from .version import check_latest, is_flagged
+    from niworkflows.utils.spaces import Reference, SpatialReferences, OutputReferencesAction
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'SpatialReferences'](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban replied:
Yes, I have the commit done, was just waiting for Circle to finish so that the build didn't get cancelled.

I have changed Codacy's settings so that only one summary comment will be posted.

oesteban commented on fmriprep/cli/workflow.py:
```diff
@@ -0,0 +1,160 @@
+    citation_files['md'].write_text(boilerplate)
+
+    if not config.execution.md_only_boilerplate and citation_files['md'].exists():
+        from subprocess import check_call, CalledProcessError, TimeoutExpired
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Consider possible security implications associated with check_call module.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented on fmriprep/tests/test_config.py:
```diff
@@ -0,0 +1,53 @@
+
+    assert [str(s) for s in spaces.get_standard(full_spec=True)] == []
+
+    assert [
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/poldracklab/fmriprep/pullRequest?prid=5112966)

oesteban commented:
> when not specifying `--output-spaces`.
> 
> Also, specifying `--output-spaces` without any arguments should not produce outputs, but it's currently not possible.

Good catch. I've addressed both use cases and added a test that doesn't cover all possibilities but is pretty close.

I've also tested this on ls5 on 15 subjects from the ABCD Study and worked like a charm.

effigies commented on fmriprep/workflows/bold/base.py:
```diff
@@ -457,8 +309,8 @@ def init_func_preproc_wf(
     # Generate a tentative boldref
-    bold_reference_wf = init_bold_reference_wf(omp_nthreads=omp_nthreads)
-    bold_reference_wf.inputs.inputnode.dummy_scans = dummy_scans
+    bold_reference_wf = init_bold_reference_wf(omp_nthreads=config.nipype.omp_nthreads)
```
Set above.

```suggestion
    bold_reference_wf = init_bold_reference_wf(omp_nthreads=omp_nthreads)
```

oesteban commented on fmriprep/workflows/tests.py:
```diff
@@ -0,0 +1,35 @@
+
+
+@contextmanager
+def mock_config():
```
Yes, it's going to be tricky, but that would be the right thing. For now, since this is only utilized in generating documentation, it seems to work well enough.

oesteban commented on fmriprep/cli/run.py:
```diff
@@ -1,745 +1,144 @@
 
-    if opts.reports_only:
-        sys.exit(int(retcode > 0))
+    config.load(config_file)
```
Okay, new comments will be up soon.

oesteban commented on fmriprep/config.py:
```diff
@@ -1,5 +1,591 @@
+    set_start_method('forkserver')
+except RuntimeError:
+    pass  # context has been already set
+finally:
```
Theoretical reason: it is explicit those are to be run regardless the branch flow went through.
Practical reason: so that my linter does not complain about imports after plain code.

Happy to take out from the finally if that feels more appropriate.

oesteban commented on fmriprep/config.py:
```diff
@@ -1,5 +1,591 @@
+    pass  # context has been already set
+finally:
+    # Defer all custom import for after initializing the forkserver and
+    # redirecting warnings
```
I meant the ignored warnings above. I'll update this comment.

oesteban commented:
I was thinking @josephmje may want to weigh in this one too.

oesteban commented:
I think this is good to go. Happy to send a quick PR to address https://github.com/poldracklab/fmriprep/pull/2018#discussion_r390417644 if it will make code clearer.

josephmje commented:
I was getting an error on [this line](https://github.com/poldracklab/fmriprep/blob/b5db031d0d3175597f69d6df64d2699b08f27f41/fmriprep/cli/parser.py#L354) when running this. It looks like `config.nipype.omp_nthread` is missing an 's'.

effigies commented:
Good catch. Care to submit a quick fix?

oesteban commented:
> Good catch. Care to submit a quick fix?

Yay :) - please f-strings :)
